openmmtools/alchemy.py

```diff
@@ -1869,261 +2104,311 @@ def _alchemically_modify_NonbondedForce(self, reference_force, alchemical_region
+                else:
+                    electrostatics_parameters = [charge, sigma]
+                for force in all_electrostatics_custom_nonbonded_forces:
+                    force.addParticle(electrostatics_parameters)
+            # Turn off interactions contribution from alchemically-modified particles in unmodified
+            # NonbondedForce that will be handled by all other forces
+            for particle_index in range(nonbonded_force.getNumParticles()):
+                # Retrieve parameters.
+                [charge, sigma, epsilon] = nonbonded_force.getParticleParameters(particle_index)
+                # Wit exact treatment of the PME electrostatics, the NonbondedForce handles the electrostatics.
+                if not use_exact_pme_treatment:
+                    charge = abs(0.0*charge)
+                if particle_index in alchemical_atomset_0:
+                    nonbonded_force.setParticleParameters(particle_index, charge, sigma, abs(0*epsilon))
+                if particle_index in alchemical_atomset_1:
+                    nonbonded_force.setParticleParameters(particle_index, charge, sigma, abs(0*epsilon))
+
+            # Restrict interaction evaluation of CustomNonbondedForces to their respective atom groups.
+            if len(region_names) > 1:
+                aa_sterics_custom_nonbonded_force.addInteractionGroup(alchemical_atomset_0, alchemical_atomset_1)
+            else:
+                na_sterics_custom_nonbonded_force.addInteractionGroup(nonalchemical_atomset, alchemical_atomset_0)
+                aa_sterics_custom_nonbonded_force.addInteractionGroup(alchemical_atomset_0, alchemical_atomset_0)
+            if use_exact_pme_treatment:
+                # The custom force only holds the original charges. It doesn't contribute to the energy.
+                original_charges_custom_nonbonded_force.addInteractionGroup({}, alchemical_atomset)
+            else:
+                if len(region_names) > 1:
+                    aa_electrostatics_custom_nonbonded_force.addInteractionGroup(alchemical_atomset_0, alchemical_atomset_1)
+                else:
+                    na_electrostatics_custom_nonbonded_force.addInteractionGroup(nonalchemical_atomset, alchemical_atomset_0)
+                    aa_electrostatics_custom_nonbonded_force.addInteractionGroup(alchemical_atomset_0, alchemical_atomset_0)
+
+            # ---------------------------------------------------------------
+            # Distribute exceptions contributions in appropriate bond forces
+            # ---------------------------------------------------------------
+
+            all_custom_nonbonded_forces = all_sterics_custom_nonbonded_forces + all_electrostatics_custom_nonbonded_forces
+            # Move all NonbondedForce exception terms for alchemically-modified particles to CustomBondForces.
+            if len(region_names) > 1:
+                for exception_index in range(reference_force.getNumExceptions()):
```

Comment:
I think the two for loops (for `len(region_names) > 1` and `else`) are identical beside for the definition of `both_alchemical` and `only_one_alchemical`, with the latter always `False` in the `len(region_names) > 1` so you could have a single for loop with an internal if-else rather than the opposite.